spellbook/optimization/qoi.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+import numpy as np
+
+from spellbook.utils import load_infile
+
+
+""" Construct a quantity of interest (cost function) for use in optimization"""
+
+
+def barrier(x, threshold, threshold_type="greater"):
+    """Implement the 'barrier penalty' function from Haftka and Starnes, AIAA Journal 14 (6): 718 (1976).
+
+    x is the quantity
+    threshold is the constraint threshold
+    threshold_type: one of "greater" or "less"
+
+    Note: this is just the penalty term, which must be combined with the objective
+    Use this with make_barrier_qoi to apply them
+
+    """
+    C = 3e-1
+    p = 1.0 / 3.0
+
+    if threshold_type == "less":
+        xx = -1.0 * x
+        threshold = -1.0 * threshold
+    elif threshold_type == "greater":
+        xx = 1.0 * x
+    else:
+        raise ValueError("threshold_type must be one of 'greater' or 'less'")
+
+    penalty = np.zeros_like(xx)
+    sign_x = np.sign(xx)
+    g0 = C * np.abs(xx) ** p
+    gi = xx - threshold
+    group1 = gi > g0
+    group2 = gi <= g0
+    group3 = xx >= threshold
+
+    penalty[group1] = sign_x[group1] * xx[group1] / gi[group1]
```

Comment:
I was confused with `sign_x`, but you are just using this as a place holder to ensure that a penalty is always positive.

I don't think you need `sign_x` and `np.abs(xx)` for g0, because when these values are negative, it means your constraint is satisfied, which should be caught in group3, when you set the penalty to 0.

I should add that I could be wrong :)

Reply:
you're right about sign_x and abs(xx), except that (negative number)** (fractional number) can throw nans, so these are in there to make the numerics nicer